iec/src/lowering/st.rs

```diff
@@ -0,0 +1,400 @@
+use crate::{
+    lowering::CompilationUnit,
+    mir::{
+        DuplicateSymbolError, Function, HasType, Location, Name, Scope,
+        ScopeRef, Symbol, Type,
+    },
+    utils::HasLocation,
+    Diagnostics,
+};
+use codespan::FileId;
+use codespan_reporting::diagnostic::{Diagnostic, Label};
+use rustmatic_structured_text as syntax;
+use shred_derive::SystemData;
+use specs::prelude::*;
+
+/// Lower a set of *Structured Text* files into a [`CompilationUnit`].
+pub fn lower<I>(
+    items: I,
+    world: &mut World,
+    diags: &mut Diagnostics,
+) -> CompilationUnit
+where
+    I: IntoIterator<Item = (FileId, syntax::File)>,
+{
+    crate::mir::register(world);
+
+    let global_scope = create_global_scope(world);
+
+    let mut lower = Lower {
+        state: world.system_data(),
+        global_scope,
+        diags,
+    };
+
+    for (id, file) in items {
+        log::debug!("Lowering file with {:?}", id);
+        lower.lower_file(id, &file);
+    }
+
+    CompilationUnit { global_scope }
+}
+
+/// Create the global [`Scope`] and populate it with various builtin types.
+fn create_global_scope(world: &mut World) -> Entity {
+    let mut scope = Scope::empty();
+
+    let int = world
+        .create_entity()
+        .with(Name(String::from("INT")))
+        .with(Type::Integer {
+            signed: true,
+            bit_width: 32,
+        })
+        .build();
+    scope
+        .add_symbol("INT".to_string(), Symbol::Type(int))
+        .unwrap();
+
+    log::debug!("Created the global scope: {:#?}", scope);
+
+    world.create_entity().with(scope).build()
+}
+
+/// Helper struct containing the various components we'll need during the
+/// lowering process.
+#[derive(SystemData)]
+struct State<'world> {
+    entities: Entities<'world>,
+    functions: WriteStorage<'world, Function>,
+    locations: WriteStorage<'world, Location>,
+    names: WriteStorage<'world, Name>,
+    scopes: WriteStorage<'world, Scope>,
+    scope_refs: WriteStorage<'world, ScopeRef>,
+    has_type: WriteStorage<'world, HasType>,
+}
+
+/// Temporary state used while lowering.
+struct Lower<'world, 'diag> {
+    state: State<'world>,
+    diags: &'diag mut Diagnostics,
+    global_scope: Entity,
+}
+
+/// Lowering functions.
+impl<'world, 'diag> Lower<'world, 'diag> {
+    fn lower_file(&mut self, id: FileId, ast: &syntax::File) {
+        for function in &ast.functions {
+            self.lower_function(id, function);
+        }
```

Comment:
other ast "results" will probably be added here, like function blocks and program